### PR TITLE
Introduced a Pre Services Update step for the registered services

### PR DESCRIPTION
**Pull request by Alexees**
Overview
InputDevices need to move prior to the FocusProvider performing all the Raycasts. This introduces this necessary step.

Changes
---
- Fixes part of #3614

### CODE REVIEW

File: Assets/MixedRealityToolkit/Providers/BaseDeviceManager.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ namespace Microsoft.MixedReality.Toolkit.Core.Providers
     /// <summary>
     /// Base Device manager to inherit from.
     /// </summary>
-    public class BaseDeviceManager : BaseExtensionService, IMixedRealityDeviceManager
+    public abstract class BaseDeviceManager : BaseExtensionService, IMixedRealityDeviceManager
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/Providers/BaseDeviceManager.cs]: Why change the base implementation to an Abstract. Shouldn't do this unless there is a concrete reason as Abstraction adds a small but noticeable overhead.

**Comment by Alexees** [comment]
@SimonDarksideJ If there's still something wrong with it you'd need to explain. I went over the scripts and reverted what was unnecessary for this.

File: Assets/MixedRealityToolkit/Providers/BaseDeviceManager.cs (hunk 1 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit/Providers/BaseDeviceManager.cs]: @SimonDarksideJ would you be so kind as to post a link or something confirming using abstract inheritance is slower than normal inheritance? I couldn't find anything about it.

Comment (Alexees) [reply Assets/MixedRealityToolkit/Providers/BaseDeviceManager.cs]: > Overall this seems sensible but needs a full round of performance validation and testing.

@SimonDarksideJ I agree here clearly. From what I can tell there are quite a couple of calls that are meant to alter the position of things. This is getting tricky when talking about let's say mouse delta, or scroll delta.
I guess someone would need to go through there and filter by what is relevant for the raycasting phase to be correct for all input devices. If that's even all there is to it.

File: Assets/MixedRealityToolkit/Providers/BaseDeviceManager.cs [hunk 2]
```diff
@@ -67,5 +67,7 @@ protected virtual IMixedRealityPointer[] RequestPointers(SystemType controllerTy
 
             return pointers.Count == 0 ? null : pointers.ToArray();
         }
+
+        public virtual void PreServiceUpdate() { }
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/Providers/BaseDeviceManager.cs]: I think something like `BeforeUpdate` or something similar?

The update is driven by Unity's game loop events anyway. They're not really specific to services.

Anyone have thoughts on a name?

Comment (Railboy) [reply Assets/MixedRealityToolkit/Providers/BaseDeviceManager.cs]: `EarlyUpdate` exists in Unity's experimental API.

**Comment by Railboy** [comment]
Based on discussion in #3614 it looks like this PR should be closed. Is that correct or am I misinterpreting everyone's conclusions?

**Comment by Alexees** [comment]
@Railboy I think so. But how it's going to be implemented remains. Will @luis-valverde-ms be doing this? or is it just switching the order of the two service lists?

**Comment by luis-valverde-ms** [comment]
@Alexees , we will eventually (within a few weeks, I hope) get to it but if it is blocking you, you're welcome to give it a try.

**Pull request closed without merging**